zkevm-circuits/src/witness/zstd/mod.rs

```diff
@@ -424,89 +420,971 @@ fn process_block_rle<F: Field>(
+
+            if n_streams > 1 {
+                let (byte_offset, rows, lstream_lens) = process_block_zstd_huffman_jump_table(
+                    src, 
+                    stream_offset, 
+                    huffman_rows.last().expect("last row should exist"),
+                    randomness
+                );
+                huffman_rows.extend_from_slice(&rows);
+            }
+
+            for idx in 0..n_streams {
+                let (byte_offset, rows, symbols) = process_block_zstd_lstream(
+                    src, 
+                    stream_offset,
+                    n_bytes,
+                    huffman_rows.last().expect("last row should exist"),
+                    randomness,
+                    idx,
+                    &huffman_codes
+                );
+                huffman_rows.extend_from_slice(&rows);
+
+                stream_offset = byte_offset;
+            }
+            
+            (stream_offset, huffman_rows)
+        },
+        _ => unreachable!("Invalid literals section BlockType")
+    };
+    witness_rows.extend_from_slice(&rows);
+
+    (bytes_offset, witness_rows)
+}
+
+fn process_block_zstd_literals_header<F: Field>(
+    src: &[u8],
+    byte_offset: usize,
+    last_row: &ZstdWitnessRow<F>,
+    randomness: Value<F>,
+) -> (usize, Vec<ZstdWitnessRow<F>>, BlockType, usize, usize, usize) {
+    let lh_bytes = src
+        .iter()
+        .skip(byte_offset)
+        .take(N_MAX_LITERAL_HEADER_BYTES)
+        .cloned()
+        .collect::<Vec<u8>>();
+
+    let literals_block_type = BlockType::from(lh_bytes[0] & 0x3);
+    let size_format = lh_bytes[0] & 3;
+
+    let [n_bits_fmt, n_bits_regen, n_bits_compressed, n_streams, n_bytes_header]: [usize; 5] = match literals_block_type {
+        BlockType::RawBlock | BlockType::RleBlock => {
+            match size_format {
+                0b00 | 0b10 => [1, 5, 0, 1, 1],
+                0b01 => [2, 12, 0, 1, 2],
+                0b11 => [2, 20, 0, 1, 3],
+                _ => unreachable!("size_format out of bound")
+            }
+        },
+        BlockType::ZstdCompressedBlock => {
+            match size_format {
+                0b00 => [2, 10, 10, 1, 3],
+                0b01 => [2, 10, 10, 4, 3],
+                0b10 => [2, 14, 14, 4, 4],
+                0b11 => [2, 18, 18, 4, 5],
+                _ => unreachable!("size_format out of bound")
+            }
+        },
+        _ => unreachable!("BlockType::Reserved unexpected or treeless literal section")
+    };
+
+    // Bits for representing regenerated_size and compressed_size
+    let sizing_bits = &lh_bytes.clone().into_iter().fold(vec![], |mut acc, b| {
+        acc.extend(value_bits_le(b));
+        acc
+    })[(2 + n_bits_fmt)..(n_bytes_header * N_BITS_PER_BYTE)];
+
+    let regen_size = le_bits_to_value(&sizing_bits[0..n_bits_regen]);
+    let compressed_size = le_bits_to_value(&sizing_bits[n_bits_regen..(n_bits_regen + n_bits_compressed)]);
```

Comment:
You could simplify this if you prefer.

```suggestion
    let [n_bytes_header, n_streams, regenerated_size, compressed_size]: [usize; 4] = match literals_block_type {
        BlockType::RawBlock | BlockType::RleBlock => {
            match size_format {
                0b00 | 0b10 => [1, 1, lh_bytes[0] >> 3, 0],
                0b01 => [2, 1, (lh_bytes[0] >> 4) + (lh_bytes[1] << 4), 0],
                0b11 => [3, 1, (lh_bytes[0] >> 4) + (lh_bytes[1] << 4) + (lh_bytes[2] << 12), 0],
                _ => unreachable!("size_format out of bound")
            }
        },
        BlockType::ZstdCompressedBlock => {
            match size_format {
                0b00 => [3, 1, (lh_bytes[0] >> 4) + ((lh_bytes[1] & 127) << 4), (lh_bytes[1] >> 6) + (lh_bytes[2] << 2)],
                0b01 => [3, 4, (lh_bytes[0] >> 4) + ((lh_bytes[1] & 127) << 4), (lh_bytes[1] >> 6) + (lh_bytes[2] << 2)],
                0b10 => [4, 4, (lh_bytes[0] >> 4) + (lh_bytes[1] << 4) + ((lh_bytes[2] & 3) << 12), (lh_bytes[2] >> 2) + (lh_bytes[3] << 6)],
                0b11 => [5, 4, (lh_bytes[0] >> 4) + (lh_bytes[1] << 4) + ((lh_bytes[2] & 127) << 12), (lh_bytes[2] >> 6) + (lh_bytes[3] << 2) + (lh_bytes[4] << 10)],
                _ => unreachable!("size_format out of bound")
            }
        },
        _ => unreachable!("BlockType::Reserved unexpected or treeless literal section")
    };
```

Well, after writing this code, I am not sure if it qualifies as "simplification". But please feel free to make your choice. I personally prefer the shift operations, but I don't mind if we collect and fold bits.

Reply:
IMO this is how we would be validating the decomposition of the literals header. I mean, we wish to find the regenerated and compressed size, which are bit accumulation calculations. However, because for all the above cases we have different offset/lengths for the bits to be accumulated, it'll be best done using another table altogether. Do you think there's a better way of achieving this?